docs/RFCS/20190318_error_handling.md

```diff
@@ -0,0 +1,2090 @@
+- `Newf("hello %s", "world")` vs `NewDepthf(1, "hello %s", "world")`
+- etc.
+
+## Handling chains of error causes
+
+### Accessing the cause
+
+The error types in the library implement the `causer` interface and Go
+2's `Wrapper` interface. It is thus possible to retrieve the layers of
+cause via the `Cause()` or `Unwrap()` methods.
+
+This comes with caveats; see the remaining sections for details.
+
+
+### Preservation of causes across the wire
+
+The library's types are protobuf-encodable and thus naturally their entire
+structure is preserved when transferred across the network.
+
+To ensure maximum convenience, some additional magic is provided
+to cover the following use case:
+
+1. a crdb/encodable error is constructed;
+2. it passes through some package which uses `errors.Wrap` (from `github.com/pkg/errors`, not the new library);
+3. the resulting error is sent across the wire.
+
+When this occurs, the library makes extra effort to convert the
+wrapper object from `github.com/pkg/errors` into a form that's
+encodable, so as to preserve all the chain of causes and the
+intermediate message prefixes added via `github.com/pkg/errors.Wrap()`.
+
+### Barrier errors
+
+When an assertion failure is encountered while handling an error, we want
+to mark the assertion failure to be an "unrecoverable error" and let it flow
+to the top level (and telemetry) without any form of further handling.
+
+This type of error must inhibit any other mechanism that inspects causes.
+
+For this purpose the library provides a special error type:
+`BarrierError`.  This contains an optional "internal cause" however
+*this cause is not visible to the `Cause()` and `Unwrap()` methods*
+(nor the other mechanisms provided below). This original cause is only
+visible:
+
+- when inspecting the error object, e.g. via `%+v` formatting;
+- when the error flows out of the system (e.g. towards a SQL client);
+- when reported to telemetry.
+
+The following functions create barrier errors:
+
+- `errors.AssertionFailed()`
+- `errors.NewAssertionErrorWithWrappedErr()`
+
+Additionally, a barrier error interacts with the pg error code as follows:
+
+- if the original pg error code was absent or was for a "normal" error,
+  the barrier error will mask the original code and replace it
+  with `XX000/CodeInternalError`.
+
+- if the original pg error code was present with the `XX` prefix, then
+  the original code is preserved. This covers the following situation:
+
+  1. a low level storage failure is encountered and generates
+     `CodeDataCorruptedError` (`XX0001`)
+  2. the storage failure trips an unexpected path in the code (an
+     assertion failure.)
+
+  In that case we wish that the assertion failure preserves the
+  serious original code.
+
+### Promotion of `XX` codes to barrier errors
+
+When a `XX` code is provided via `Wrapx`, `WithDefaultCode`, etc, the
+result becomes a barrier and the original error, if any, is masked as an
```

Comment:
This seems "magic" and therefore best avoided. Don't you think it would be better to force explicit construction of barrier errors?